### ISSUE TITLE
App freezes on update

### ISSUE THREAD

**Issue by viking2009** [Microsoft/cordova-plugin-code-push#156]
Please check this log. Real deployment key was replaced for security issue.

> Jul 18 18:31:50 <Warning>: [CodePush] Checking for update.
> Jul 18 18:31:51 <Warning>: [CodePush] An update is available. {"appVersion":"24.0.1","deploymentKey":"DEPLOYMENT_KEY","description":"","downloadUrl":"https://codepush.blob.core.windows.net/storagev2/BLOB","isMandatory":false,"label":"v2","packageHash":"fb6addc864445059957f6b8b11d5301e12a31ab2971924839d7d17a652210078","packageSize":7528307,"failedInstall":false}
> Jul 18 18:31:51 <Warning>: [CodePush] Downloading update
> Jul 18 18:32:37 <Warning>: File Transfer Finished with response code 200
> Jul 18 18:32:37 <Warning>: [CodePush] Package download success: {"deploymentKey":"DEPLOYMENT_KEY","description":"","label":"v2","appVersion":"24.0.1","isMandatory":false,"packageHash":"fb6addc864445059957f6b8b11d5301e12a31ab2971924839d7d17a652210078","isFirstRun":false,"failedInstall":false,"localPath":"cdvfile://localhost/library-nosync/codepush/download/update.zip"}
> Jul 18 18:32:37 <Warning>: [CodePush] Installing update
> Jul 18 18:32:45 <Warning>: THREAD WARNING: ['File'] took '4534.373047' ms. Plugin should use a background thread.
> Jul 18 18:32:46 <Warning>: [CodePush] Install succeeded.
> Jul 18 18:32:46 <Warning>: [CodePush] Update is installed and will be run on the next app restart.

on this stage
> THREAD WARNING: ['File'] took '4534.373047' ms. Plugin should use a background thread.

app freezes and doesn't respond to any actions.
The size of the file is ~7.5MB. Is it unzip stage? How we can solve the problem?

**Comment by geof90** [comment]
Which version of `cordova-plugin-file` are you using?

**Comment by viking2009** [comment]
code-push 1.9.0-beta "CodePushAcquisition"
cordova-plugin-code-push 1.8.2-beta "CodePush"

**Comment by geof90** [comment]
The problem seems to be coming from cordova-plugin-file, which is one of our dependencies. I would file an issue on them.

**Comment by viking2009** [comment]
OK. Thanks.
Could you preview ETA for this?

**Comment by silhouettes** [comment]
@viking2009, could you check which version of `cordova-plugin-file` you are running? It is possible that writing files in the background was fixed in `1.3.2`: https://github.com/apache/cordova-plugin-file/blob/d9b90461f1d0dda9c5f2337720f3333f17b89b0e/RELEASENOTES.md#132-dec-02-2014. If you upgrade to the latest version of `cordova-plugin-file`, does that resolve the issue for you?

**Comment by viking2009** [comment]
@silhouettes Thanks. We have cordova-plugin-file 4.2.0 "File"

The problem is here

```
- (void)doCopyMove:(CDVInvokedUrlCommand*)command isCopy:(BOOL)bCopy
{
    //...
    __weak CDVFile* weakSelf = self;
    [destFs copyFileToURL:destURL withName:newName fromFileSystem:srcFs atURL:srcURL copy:bCopy callback:^(CDVPluginResult* result) {
        [weakSelf.commandDelegate sendPluginResult:result callbackId:command.callbackId];
    }];

}
```

```callback``` is executed in main thread

> __29-[CDVFile doCopyMove:isCopy:]_block_invoke:isMainThread:1
> __29-[CDVFile doCopyMove:isCopy:]_block_invoke:4664.228027
> THREAD WARNING: ['File'] took '4664.548828' ms. Plugin should use a background thread.

I fixed this with

```
- (void)doCopyMove:(CDVInvokedUrlCommand*)command isCopy:(BOOL)bCopy
{
    //...
    __weak CDVFile* weakSelf = self;
     [self.commandDelegate runInBackground:^ {
        [destFs copyFileToURL:destURL withName:newName fromFileSystem:srcFs atURL:srcURL copy:bCopy callback:^(CDVPluginResult* result) {
            [weakSelf.commandDelegate sendPluginResult:result callbackId:command.callbackId];
        }];
    }];
}
```

> __29-[CDVFile doCopyMove:isCopy:]_block_invoke_2:isMainThread:0
> __29-[CDVFile doCopyMove:isCopy:]_block_invoke_2:4492.204102

I can't add issue to https://github.com/apache/cordova-plugin-file

**Comment by silhouettes** [comment]
Thanks so much for the info @viking2009, this is very useful. I've filed an issue on their JIRA tracker here: https://issues.apache.org/jira/browse/CB-11693. Feel free to come back if you have any other issues!

**Comment by viking2009** [comment]
Hi.
There is a pull request https://github.com/apache/cordova-plugin-file/pull/190.
Do you plan to update cordova-plugin-file version as soon as the pull request will be merged?
Thanks.

**Comment by silhouettes** [comment]
Yep, we can update the dependency when the new version is released!

**Comment by galsch** [comment]
@silhouettes Hi, could you please provide ETA when the new version is released(maybe you could ask cordova-plugin-file when they will release their new fixed version and then give an additional estimate you will also be ready with an new version)?

We are really excited and looking forward to integrate  Codepush into our product to production. However, this issue blocks us from doing so.

It would be great if you could let's know once you have the release dates for the new plugin with the fix - this way we can prepare our QA/Dev resources.

Thanks

**Comment by galsch** [comment]
@silhouettes any news?

**Comment by silhouettes** [comment]
Hi @galsch, I'm unable to influence when cordova-plugin-file makes a new release, perhaps you could try to ask them? As soon as they make the release, we will be able to update our plugin within one or two days.

In the meantime, you should be able to just clone/fork the corresponding repos and use that as a workaround.

**Comment by silhouettes** [comment]
Perhaps file an issue with them here, similarly to what I did previously: https://issues.apache.org/jira/browse/CB/?selectedTab=com.atlassian.jira.jira-projects-plugin:summary-panel

**Comment by galsch** [comment]
@silhouettes hi and thanks for the reply.

If Microsoft is unable to get in touch with one of its dependencies in their main product - we are in serious problem :)

If CodePush is not something we can rely on in production then we will have to switch to a different product/solution.

Of course we won't fork/clone something which is unstable until an actual public release is available.

I suggest that you take closer look in the team to understand how it's possible to have such a critical(IMHO) bug unresolved for so long - eventually the bug is in your product (it doesn't matter to the clients if it's a third party issue - we use your product).

Thanks

**Comment by silhouettes** [comment]
Hi @galsch, we are an open source product with limited resources. We are unable to address every single request, and as a result we must prioritize issues accordingly. Because of this, we simply ask for patience as issues are addressed, and where possible, for assistance.

We have reached out internally to `cordova-plugin-file`, but keep in mind this is a core plugin, and a separate project with its own priorities. As mentioned, once they perform a release, we will also perform one. As far as CodePush, I believe we have done everything we can reasonably do to help you, given the current prioritization of this issue.

I want to reiterate once again that this is an open source project that is free to use. We expect to receive co-operation, if not contribution, as issues are resolved. I believe we have provided multiple avenues for addressing this issue, including cloning/forking the relevant repo, which is a commonly accepted way in the open source community of addressing co-ordination problems of this nature.

**Comment by silhouettes** [comment]
@galsch, @viking2009: cordova-plugin-file has just been released, and we have updated the dependency in master. We will do some testing over the next two days before making a release, possibly with other fixes.

**Comment by silhouettes** [comment]
My sincere apologies for the slow response on this - this has now been published in `1.9.0-beta` ([release notes](https://github.com/Microsoft/cordova-plugin-code-push/releases/tag/v1.9.0-beta)). Thank you so much for your patience, and please feel free to report any future issues and we will follow up on them as soon as we can.
